pkg/metrics/instance/configstore/remote.go

```diff
@@ -347,111 +268,32 @@ func (r *Remote) all(ctx context.Context, keep func(key string) bool) (<-chan in
-	//	configs this overloaded the consul instances. This reduces that to one call, that was being made anyways.
-	if r.kv.consul != nil {
-		return r.allConsul(ctx, keep)
-	}
-	return r.allOther(ctx, keep)
-
-}
-
-// allConsul is ONLY usable when consul is the keystore. This is a performance improvement in using the client directly
-//	instead of the cortex multi store kv interface. That interface returns the list then each value must be retrieved
-//	individually. This returns all the keys and values in one call and works on them in memory
-func (r *Remote) allConsul(ctx context.Context, keep func(key string) bool) (<-chan instance.Config, error) {
-	if r.kv.consul == nil {
-		level.Error(r.log).Log("err", "allConsul called but consul client nil")
-		return nil, errors.New("allConsul called but consul client nil")
+	kvps, err := r.kv.List(ctx, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to list configs: %w", err)
 	}
-	var configs []*instance.Config
-	c := GetCodec()
 
-	pairs, err := r.listConsul(ctx)
+	// NOTE(rfratto): ch must be buffered for everything we're going to write to
+	// it, otherwise this goroutine will deadlock.
+	//
+	// Data written to the buffered channel will still be available for reading
+	// even after the channel is closed.
+	ch := make(chan instance.Config, len(kvps))
+	defer close(ch)
 
-	if err != nil {
-		return nil, err
-	}
-	for _, kvp := range pairs {
+	for _, kvp := range kvps {
 		if keep != nil && !keep(kvp.Key) {
 			level.Debug(r.log).Log("msg", "skipping key that was filtered out", "key", kvp.Key)
 			continue
 		}
-		value, err := c.Decode(kvp.Value)
-		if err != nil {
-			level.Error(r.log).Log("msg", "failed to decode config from store", "key", kvp.Key, "err", err)
-			continue
-		}
-		if value == nil {
-			// Config was deleted since we called list, skip it.
-			level.Debug(r.log).Log("msg", "skipping key that was deleted after list was called", "key", kvp.Key)
-			continue
-		}
 
-		cfg, err := instance.UnmarshalConfig(strings.NewReader(value.(string)))
+		cfg, err := instance.UnmarshalConfig(strings.NewReader(kvp.Value.(string)))
 		if err != nil {
 			level.Error(r.log).Log("msg", "failed to unmarshal config from store", "key", kvp.Key, "err", err)
 			continue
 		}
-		configs = append(configs, cfg)
-	}
-	ch := make(chan instance.Config, len(configs))
-	for _, cfg := range configs {
 		ch <- *cfg
 	}
-	close(ch)
-	return ch, nil
-}
-
-func (r *Remote) allOther(ctx context.Context, keep func(key string) bool) (<-chan instance.Config, error) {
-	if r.kv == nil {
-		return nil, ErrNotConnected
-	}
-
-	keys, err := r.kv.List(ctx, "")
-	if err != nil {
-		return nil, fmt.Errorf("failed to list configs: %w", err)
-	}
-
-	ch := make(chan instance.Config)
-
-	var wg sync.WaitGroup
-	wg.Add(len(keys))
-	go func() {
-		wg.Wait()
-		close(ch)
-	}()
-
-	for _, key := range keys {
-		go func(key string) {
```

Comment:
@mattdurham I decided that your approach of iterating over all the keys in a single goroutine and then flushing to a channel will likely be more performant, especially if there's thousands of keys.